### PR TITLE
Add new filter to parse xml output for network use cases

### LINKED ISSUE #31026: Add parse xml filter which returns a json object [XML_FILTER]
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

filter/network.py

##### ANSIBLE VERSION

```
2.5
```

##### SUMMARY

*  Add parse_xml filter to parse a xml response and convert to json object.

##### STEPS TO REPRODUCE

```yaml

```

##### ACTUAL RESULTS

```

```

### PR DESCRIPTION
**Pull request by ganeshrn**
##### SUMMARY

Fixes #31026
*  Add parse_xml filter
*  Add documentation for parse_xml filter

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

plugins/filter/network.py

##### ANSIBLE VERSION

```
2.5
```

##### ADDITIONAL INFORMATION

Playbook:
```
---
- name: parse xml output
  hosts: localhost
  vars:
    output: >
            <rpc-reply>
              <configuration>
                <vlans>
                  <vlan inactive="inactive">
                   <name>vlan-1</name>
                   <vlan-id>100</vlan-id>
                   <description>This is vlan-1</description>
                  </vlan>
                </vlans>
              </configuration>
            </rpc-reply>

  tasks:
  - name: filter xml output
    debug:
      msg: "{{ output | parse_xml('vlan_spec_xml.yml') }}"
```
Spec file:
```
#cat vlan_spec_xml.yml
---
vars:
  vlan:
    vlan_id: "{{ item.vlan_id }}"
    name: "{{ item.name }}"
    desc: "{{ item.desc }}"
    enabled: "{{ item.state.get('inactive') != 'inactive' }}"
    state: "{% if item.state.get('inactive') == 'inactive'%}inactive{% else %}active{% endif %}"

keys:
  vlans:
    type: list
    value: "{{ vlan }}"
    top: configuration/vlans/vlan
    items:
      vlan_id: vlan-id
      name: name
      desc: description
      state: ".[@inactive='inactive']"
```

Output:
```
TASK [filter xml output] ************************************************************************************
task path: /home/users/parse_xml_filter_vlan.yml:18
ok: [localhost] => {
    "msg": {
        "vlans": [
            {
                "desc": "This is vlan-1",
                "enabled": false,
                "name": "vlan-1",
                "state": "inactive",
                "vlan_id": 100
            }
        ]
    }
}
```

### CODE REVIEW

File: docs/docsite/rst/playbooks_filters.rst [hunk 1]
```diff
@@ -421,6 +421,90 @@ filter::
...
+    keys:
+      vlans:
+        type: list
+        value: "{{ vlan }}"
+        top: configuration/vlans/vlan
+        items:
+          vlan_id: vlan-id
+          name: name
+          desc: description
+          state: ".[@inactive='inactive']"
```
Comment (privateip) [inline docs/docsite/rst/playbooks_filters.rst:485]: probably need to explain this a little more (usage of `.[@inactive='inactive']`)

**Review by privateip — CHANGES_REQUESTED** [review]
overall it looks good.  the only comment would be to add some notes around how to extract xml tags

File: docs/docsite/rst/playbooks_filters.rst [hunk 2]
```diff
@@ -421,6 +421,90 @@ filter::
...
+
+.. versionadded:: 2.5
+
+To convert the XML output of a network device command into structured JSON
+output, use the ``parse_xml`` filter::
+
+  {{ output | parse_xml('path/to/spec') }}
+
+The ``parse_xml`` filter will load the spec file and pass the command output
+through, it returning JSON output. The spec file is a YAML yaml that defines
```
Comment (gundalow) [inline docs/docsite/rst/playbooks_filters.rst]: Should this read
> The spec file is a YAML file that defines

**Review by gundalow — CHANGES_REQUESTED** [review]
Would like to see some unit tests adding for this

**Comment by gundalow** [comment]
Once we are happy with the technical content it will need final review from @dharmabumstead

File: docs/docsite/rst/playbooks_filters.rst [hunk 3]
```diff
@@ -421,6 +421,91 @@ filter::
...
+        value: "{{ vlan }}"
+        top: configuration/vlans/vlan
+        items:
+          vlan_id: vlan-id
+          name: name
+          desc: description
+          state: ".[@inactive='inactive']"
+
+
+The example xml output from a network device shown below is used to identify the value of ``top`` and
```
Comment (dharmabumstead) [inline docs/docsite/rst/playbooks_filters.rst]: Pleae revise this paragraph - it’s difficult to parse.

**Review by dharmabumstead — CHANGES_REQUESTED** [review]

File: docs/docsite/rst/playbooks_filters.rst (hunk 1 above)
Comment (ganeshrn) [reply docs/docsite/rst/playbooks_filters.rst:485]: Fixed

**Comment by ganeshrn** [comment]
ready_for_review

**Comment by ganeshrn** [comment]
ready_for_review

File: docs/docsite/rst/playbooks_filters.rst [hunk 4]
```diff
@@ -421,6 +419,99 @@ filter::
...
+        value: "{{ vlan }}"
+        top: configuration/vlans/vlan
+        items:
+          vlan_id: vlan-id
+          name: name
+          desc: description
+          state: ".[@inactive='inactive']"
+
+
+The value of ``top`` is the relatvie XPath till inner-most container in xml heirarchy.
```
Comment (gundalow) [inline docs/docsite/rst/playbooks_filters.rst]: Typo relative

File: docs/docsite/rst/playbooks_filters.rst [hunk 5]
```diff
@@ -421,6 +419,99 @@ filter::
...
+From the example xml output given below the value of ``top`` is ``configuration/vlans/vlan`` which
+is a relative XPath expression wrt to root node (ie. rpc-reply)
```
Comment (gundalow) [inline docs/docsite/rst/playbooks_filters.rst]: Please replace wrt: with respect to

**Review by gundalow — CHANGES_REQUESTED** [review]

**Review by gundalow — APPROVED** [review]

**Review by privateip — APPROVED** [review]

File: docs/docsite/rst/playbooks_filters.rst (hunk 4 above)
Comment (dharmabumstead) [reply docs/docsite/rst/playbooks_filters.rst]: Replace typo “relatvie” with “relative”
Fix “till inner-most container”

File: docs/docsite/rst/playbooks_filters.rst (hunk 5 above)
Comment (dharmabumstead) [reply docs/docsite/rst/playbooks_filters.rst]: Capitalize “XML”

File: docs/docsite/rst/playbooks_filters.rst [hunk 6]
```diff
@@ -340,7 +340,7 @@ output, use the ``parse_cli`` filter::
   {{ output | parse_cli('path/to/spec') }}
 
 The ``parse_cli`` filter will load the spec file and pass the command output
-through, it returning JSON output.  The spec file is a YAML yaml that defines
+through, it returning JSON output.  The spec file is a YAML that defines
```
Comment (dharmabumstead) [inline docs/docsite/rst/playbooks_filters.rst]: Please revise to “The YAML spec file defines...”

File: docs/docsite/rst/playbooks_filters.rst [hunk 7]
```diff
@@ -421,6 +419,99 @@ filter::
...
+        value: "{{ vlan }}"
+        top: configuration/vlans/vlan
+        items:
+          vlan_id: vlan-id
+          name: name
+          desc: description
+          state: ".[@inactive='inactive']"
+
+
+The value of ``top`` is the relative XPath till inner-most container in xml heirarchy.
```
Comment (dharmabumstead) [inline docs/docsite/rst/playbooks_filters.rst]: Hard to parse this sentence. Please rephrase. 

Please capitalize XML.

File: docs/docsite/rst/playbooks_filters.rst [hunk 8]
```diff
@@ -421,6 +419,99 @@ filter::
...
+From the example xml output given below the value of ``top`` is ``configuration/vlans/vlan`` which
```
Comment (dharmabumstead) [inline docs/docsite/rst/playbooks_filters.rst]: Please capitalize XML. Please use definite articles (“the”).

File: docs/docsite/rst/playbooks_filters.rst [hunk 9]
```diff
@@ -421,6 +419,99 @@ filter::
...
+is a relative XPath expression with respect to root node (ie. rpc-reply)
```
Comment (dharmabumstead) [inline docs/docsite/rst/playbooks_filters.rst]: Please use “for exsmpleL instead of “i.e.”

File: docs/docsite/rst/playbooks_filters.rst [hunk 10]
```diff
@@ -421,6 +419,99 @@ filter::
...
+
+``items`` is a dictionary, of key-value pairs that map user-defined names to XPath expressions
+that select elements. The Xpath expression is relative XPath with respect to value of XPath in ``top``.
+For example the ``vlan_id`` in spec file is user defined name and it's value ``vlan-id`` is the
+relative XPath with respect to value of XPath in ``top``
+
+Attributes of xml tags can be extracted using XPath expressions, the value of ``state`` in spec
```
Comment (dharmabumstead) [inline docs/docsite/rst/playbooks_filters.rst]: Please capitalize “XML” throughout.

**Review by dharmabumstead — CHANGES_REQUESTED** [review]
Some edits required. Thanks @ganeshrn!

**Comment by ganeshrn** [comment]
ready_for_review

**Comment by ganeshrn** [comment]
ready_for_review

**Review by dharmabumstead — APPROVED** [review]
Docs OK with a few more copy edits. OK to merge if the networking team has tech reviewed and approved. Thanks @ganeshrn!

**Comment by ganeshrn** [comment]
@privateip @gundalow @dharmabumstead Thank you for reviewing.
